nexus/reconfigurator/planning/src/planner/omicron_zone_placement.rs

```diff
@@ -0,0 +1,493 @@
+            | BlueprintZoneType::ClickhouseKeeper(_)
+            | BlueprintZoneType::CockroachDb(_)
+            | BlueprintZoneType::CruciblePantry(_)
+            | BlueprintZoneType::ExternalDns(_)
+            | BlueprintZoneType::InternalDns(_)
+            | BlueprintZoneType::Oximeter(_) => None,
+            // Zones that get special handling for placement (all sleds get
+            // them, although internal NTP has some interactions with boundary
+            // NTP that we don't yet handle, so this may change).
+            BlueprintZoneType::Crucible(_)
+            | BlueprintZoneType::InternalNtp(_) => None,
+        }
+    }
+}
+
+impl From<DiscretionaryOmicronZone> for ZoneKind {
+    fn from(zone: DiscretionaryOmicronZone) -> Self {
+        match zone {
+            DiscretionaryOmicronZone::Nexus => Self::Nexus,
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub(super) enum PlacementError {
+    #[error(
+        "no sleds eligible for placement of new {} zone",
+        ZoneKind::from(*zone_kind)
+    )]
+    NoSledsEligible { zone_kind: DiscretionaryOmicronZone },
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub(super) struct OmicronZonePlacementSledState {
+    pub sled_id: SledUuid,
+    pub num_zpools: usize,
+    pub discretionary_zones: Vec<DiscretionaryOmicronZone>,
+}
+
+/// `OmicronZonePlacement` keeps an internal heap of sleds and their current
+/// discretionary zones and chooses sleds for placement of additional
+/// discretionary zones.
+#[derive(Debug, Clone)]
+pub(super) struct OmicronZonePlacement {
+    sleds: OrderedSleds,
+}
+
+impl OmicronZonePlacement {
+    /// Construct a new `OmicronZonePlacement` with a given set of eligible
+    /// sleds.
+    ///
+    /// Sleds which are not eligible for discretionary services for reasons
+    /// outside the knowledge of `OmicronZonePlacement` (e.g., sleds with a
+    /// policy or state that makes them ineligible) should be omitted from this
+    /// list of sleds. For now, sleds that are waiting for an NTP zone should be
+    /// omitted as well, although that may change in the future when we add
+    /// support for boundary NTP zone placement.
+    pub(super) fn new(
+        sleds: impl Iterator<Item = OmicronZonePlacementSledState>,
+    ) -> Self {
+        // We rebuild our heap whenever the zone type we're placing changes. We
+        // need to pick _something_ to start; this only matters for performance,
+        // not correctness (we don't have to rebuild the heap if `place_zone` is
+        // called with a zone kind that matches the current sorting).
+        let ordered_by = DiscretionaryOmicronZone::Nexus;
+        Self { sleds: OrderedSleds::new(ordered_by, sleds) }
+    }
+
+    /// Attempt to place a new zone of kind `zone_kind` on one of the sleds
+    /// provided when this `OmicronZonePlacement` was created.
+    ///
+    /// On success, the internal heap held by `self` is updated assuming that a
```

Comment:
They definitely should accept the decision! I just wanted to document an edge case in usage, similar to the comment we have on `BlueprintZonesBuilder::change_sled_zones()`: https://github.com/oxidecomputer/omicron/blob/a5213df36ed20f5ee8a63e46fe0f277460950158/nexus/reconfigurator/planning/src/blueprint_builder/builder.rs#L871-L874; i.e., "if you call this function, you need to actually use the result in a way that we don't check".

I think it would be better if neither of these needed this kind of comment, but I'm not sure how to rework them to avoid it.